pkg/costmodel/allocation_helpers.go

```diff
@@ -0,0 +1,133 @@
+package costmodel
+
+import (
+	"sort"
+	"time"
+
+	"github.com/kubecost/cost-model/pkg/kubecost"
+)
+
+// IntervalPoint describes a start or end of a window of time
+// Currently, this used in PVC-pod relations to detect/calculate
+// coefficients for PV cost when a PVC is shared between pods.
+type IntervalPoint struct {
+	Time      time.Time
+	PointType string
+	Key       podKey
+}
+
+// NewIntervalPoint creates and returns a new IntervalPoint instance with given parameters.
+func NewIntervalPoint(time time.Time, pointType string, key podKey) IntervalPoint {
+	return IntervalPoint{
+		Time:      time,
+		PointType: pointType,
+		Key:       key,
+	}
+}
+
+// getIntervalPointFromWindows takes a map of podKeys to windows
+// and returns a sorted list of IntervalPoints representing the
+// starts and ends of all those windows.
+func getIntervalPointsFromWindows(windows map[podKey]kubecost.Window) []IntervalPoint {
+
+	var intervals []IntervalPoint
+
+	for podKey, podInterval := range windows {
+
+		start := NewIntervalPoint(*podInterval.Start(), "start", podKey)
+		end := NewIntervalPoint(*podInterval.End(), "end", podKey)
+
+		intervals = append(intervals, []IntervalPoint{start, end}...)
+
+	}
+
+	sortIntervalPoints(intervals)
+
+	return intervals
+
+}
+
+// sortIntervalPoints sorts a list of IntervalPoints from earliest
+// to latest IntervalPoint.Time. In the case that two IntervalPoints
+// have the same time, the point with Type "start" is treated as coming
+// before the "end".
+func sortIntervalPoints(intervals []IntervalPoint) {
+	sort.Slice(intervals, func(i, j int) bool {
+		if intervals[i].Time.Equal(intervals[j].Time) {
+			return intervals[i].PointType == "start" && intervals[j].PointType == "end"
+		}
+		return intervals[i].Time.Before(intervals[j].Time)
+	})
+}
+
+// getPVCCostCoefficients gets a coefficient which represents the scale
+// factor that each PVC in a pvcIntervalMap and corresponding slice of
+// IntervalPoints intervals uses to calculate a cost for that PVC's PV.
+func getPVCCostCoefficients(intervals []IntervalPoint, pvcIntervalMap map[podKey]kubecost.Window, pvcCostCoefficientMap map[podKey][][]float64) {
+
+	// pvcCostCoefficientMap is mutated in this function. The format is
+	// such that the individual coefficient components are preserved for
+	// testing purposes.
+
+	var activePods float64
+
+	activeKeys := make(map[podKey]struct{})
+
+	// For each interval i.e. for any time a pod-PVC relation ends or starts...
+	for i := 1; i < len(intervals); i++ {
+
+		point := intervals[i]
+		prevPoint := intervals[i-1]
+
+		// intervals will always have at least two IntervalPoints (one start/end)
+		if i == 1 {
+			activePods += 1
+			activeKeys[prevPoint.Key] = struct{}{}
+		}
```

Comment:
Is this meant to be the initialization? If so, what do you think about doing this outside the `for` loop so that we don't check this condition each time?

Something like this, above the loop:
```go
// Track active state, starting after the first point (i.e. first pod) has been recorded.

// Number of currently active pods, tracked through the sweep
activePods := 1.0

// Set of keys from all points that have reached "start" but not yet "end"
activeKeys := map[podKey]struct{}{
  intervals[0].prevPoint.Key: struct{}{},
}
```

Reply:
This is a little hacky, but essentially it's adding a map entry with a zero memory size value just to act as a sort of "record" for what specific pods are active at that point in the loop, because each of those pods must then [get a piece of their coefficient appended](https://github.com/kubecost/cost-model/pull/1036/files/5328ab40e1868727b00c7237a913e54f411d3d2b#diff-fb9d5089e2629c9bd167b04b7a80f1d70db00e805dabeed463be0913fd101e8aR90) to their key in the result map.

This could be done with a slice, but since elements need to be deleted out of order, I used a map.

Reply:
Right, totally with you on the purpose. My question was meant to be about how you're initializing it. As written, this checks for initialization every single time through the loop. Proposal was only to make the initial condition explicit above the loop so we don't have to check `if i == 1` repeatedly.

Reply:
Ack, you're absolutely correct, I wasn't looking at the right part of the code.

In that case, totally agree.

Reply:
Implemented this.